### PR TITLE
[Core] Improve AbstractRector: ensure passed $mutatingScope arg on ChangedNodeScopeRefresher::refresh() is MutatingScope instance

### PR DESCRIPTION
**Pull request by samsonasik**
ensure passed $mutatingScope arg on ChangedNodeScopeRefresher::refresh() is MutatingScope instance by:

- check early Nodes that not have Scope, 
- check scope instanceof MutatingScope
- check not has unreacable statement

### CODE REVIEW

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by samsonasik** [comment]
@TomasVotruba after some thinking, I think `UnreachableStmtAnalyzer` actually not needed. The `Scope` require refresh ShouldNotHappenException can happen when Node is reachable, but current `Stmt` is lost (null).

I updated the check to avoid loop previous node that can reduce performance, as if node already unreacable, last stmt is already unreacable as well.

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready 🎉 🎉 🎉 

I tested build script locally, and it is working ok : 

Screen Shot 2022-05-12 at 19 55 51

File: src/Rector/AbstractRector.php [hunk 1]
```diff
@@ -251,58 +248,40 @@ final public function enterNode(Node $node)
...
-            if ($requiresScopeRefresh) {
-                $this->changedNodeScopeRefresher->refresh($node, $this->file->getSmartFileInfo(), $currentScope);
-            }
+        if (in_array(
+            $node::class,
+            [Name::class, FullyQualified::class, Namespace_::class, FileWithoutNamespace::class, Identifier::class],
+            true
+        )) {
+            return $this->processResultSingleNode($node, $originalNode);
         }
```
Comment (TomasVotruba) [inline src/Rector/AbstractRector.php]: These should be `instanceof`, as we use this convention to check every node type here. The `::class` takes only current one and requires repeating the whole tree.

File: src/Rector/AbstractRector.php [hunk 2]
```diff
@@ -251,58 +248,40 @@ final public function enterNode(Node $node)
...
+            $this->changedNodeScopeRefresher->refresh($node, $this->file->getSmartFileInfo(), $currentScope);
+            return $this->processResultSingleNode($node, $originalNode);
         }
 
-        // search "infinite recursion" in https://github.com/nikic/PHP-Parser/blob/master/doc/component/Walking_the_AST.markdown
-        $originalNodeHash = spl_object_hash($originalNode);
+        $isUnreachable = $node->getAttribute(AttributeKey::IS_UNREACHABLE);
+        if ($isUnreachable === true) {
+            return $this->processResultSingleNode($node, $originalNode);
+        }
```
Comment (TomasVotruba) [inline src/Rector/AbstractRector.php]: The unreachable logic should be kept outside the service, as we'll use it in another places too.
Also the `AbstractRector` is pretty huge now, so the aim is to make it as thin as possible.

File: src/Rector/AbstractRector.php (hunk 1 above)
Comment (samsonasik) [reply src/Rector/AbstractRector.php]: ok, use instanceof check back

File: src/Rector/AbstractRector.php (hunk 2 above)
Comment (samsonasik) [reply src/Rector/AbstractRector.php]: ok, re-add `UnreachableStmtAnalyzer`, with look by current stmt, directly check IS_UNREACHABLE, fallback to true when current stmt is not Stmt.

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by samsonasik** [comment]
I am closing for now, I tried to get is unreachable attribute, but pull from node instead of Stmt seems not quite right, while working.

Loop previous node, on previous implementation seems not ideal as well, and I think needs to be refactored when possible.

**Pull request closed without merging**